consensus/src/experimental/commit_phase.rs

```diff
@@ -190,73 +144,52 @@ impl CommitPhase {
         &mut self,
         commit_decision: CommitDecision,
     ) -> anyhow::Result<()> {
-        let li = commit_decision.ledger_info();
+        if let Some((vecblock, leder_info)) = self.blocks.as_ref() {
+            let li = commit_decision.ledger_info();
 
-        if li.ledger_info().commit_info().round() < self.committed_round {
-            return Ok(()); // we ignore the message
+            // if the block infos do not match
+            if !(li.ledger_info().commit_info().clone()
+                == leder_info.ledger_info().commit_info().clone())
+            {
+                return Ok(()); // ignore the message
+            }
+
+            commit_decision.verify(&self.verifier)?;
+
+            // replace the signature tree
+            self.blocks = Some((vecblock.clone(), li.clone()));
         }
 
-        commit_decision.verify(&self.verifier)?;
+        Ok(())
+    }
 
-        let executed_state_hash = li.ledger_info().commit_info().executed_state_id();
+    pub async fn process_block(&mut self) -> anyhow::Result<()> {
+        if let Some((vecblock, leder_info)) = self.blocks.as_ref() {
+            if leder_info.verify_signatures(&self.verifier).is_ok() {
+                commit(&self.execution_proxy, &vecblock, leder_info)
+                    .await
+                    .expect("Failed to commit the executed blocks.");
 
-        // TODO: optimization1: probe local_cache first to see if the existing already verifies,
-        // TODO: otherwise we do not make changes.
-        // TODO: optimization2: we can set a bit to indicate this tree of signatures are already verified,
-        // TODO: we do not have to verify it again in the main loop.
+                let commit_round = vecblock.last().unwrap().block_info().round();
```

Comment:
if we're using the PendingBlock type, we can have function to get the last block info to avoid writing last().unwrap().block_info()

Reply:
resolved